base/channels.jl

```diff
@@ -0,0 +1,137 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+abstract AbstractChannel{T}
+
+type Channel{T} <: AbstractChannel{T}
+    cond_take::Condition    # waiting for data to become available
+    cond_put::Condition     # waiting for a writeable slot
+    state::Symbol
+
+    data::Array{T,1}
+    szp1::Int               # current channel size plus one
+    sz_max::Int             # maximum size of channel
+    take_pos::Int           # read position
+    put_pos::Int            # write position
+
+    function Channel(sz)
```

Comment:
A further optimization that should be done is halving `data` once the current number of objects is less than 50% of `data` length